### PR TITLE
types: clean up JSON unmarshaling story for extension types

### PR DESCRIPTION
**Pull request by patjakdev**
*Issue #, if available:* None

*Description of changes:*

Create a generic `unmarshalExtensionValue()` helper function that supports all three forms of extension value JSON encoding:

* Explicit: `{ "__extn": { "fn": "decimal", "arg": "1234.5678" } }`
* Implicit object: `{ "fn": "decimal", "arg": "1234.5678" } }`
* Implicit string: `"1234.5678"`

Use it to flesh out the implementations of `UnmarshalJSON` for `IPAddr` and `Decimal`. Note that the implicit decoding will never actually be invoked until we have schema-driven decoding for entities.

### CODE REVIEW

File: types/json.go [hunk 1]
```diff
@@ -125,3 +126,42 @@ func UnmarshalJSON(b []byte, v *Value) error {
 	}
 	return nil
 }
+
+func unmarshalExtensionValue[T any](b []byte, extName string, parse func(string) (T, error)) (T, error) {
+	var zeroT T
+	var arg string
```
Comment (apg) [inline types/json.go:132]: nit: you could combine this into one var declaration

```suggestion
	var (
		zeroT T
		arg string
	)
```

Comment (patjakdev) [reply types/json.go:132]: I think I prefer the two line version...

File: types/json.go [hunk 2]
```diff
@@ -125,3 +126,42 @@ func UnmarshalJSON(b []byte, v *Value) error {
...
+		}
+	} else {
+		var res extValueJSON
+		if err := json.Unmarshal(b, &res); err != nil {
+			return zeroT, errors.Join(errJSONDecode, err)
+		}
+		if res.Extn == nil {
+			// If we didn't find an Extn, maybe it's just an extn.
+			var res2 extn
+			_ = json.Unmarshal(b, &res2)
```
Comment (patjakdev) [inline types/json.go:145]: Note to self: document why it's okay to ignore this error...

Comment (philhassey) [reply types/json.go:149]: Or not!

**Review by philhassey — APPROVED** [review]
LGTM

**Pull request merged**